### PR TITLE
feat(email-service): Refactor code and improve logic based on feedback

### PR DESCRIPTION
**Pull request by KShivendu**
## Summary of change

Changes suggested by @bhumilsarvaiya in the main PR.

## Related PR

- #100

### CODE REVIEW

File: supertokens_python/recipe/emailverification/recipe.py [hunk 1]
```diff
@@ -51,7 +53,8 @@ class EmailVerificationRecipe(RecipeModule):
 
     def __init__(self, recipe_id: str, app_info: AppInfo,
                  config: ParentRecipeEmailVerificationConfig,
-                 email_delivery_ingredient: Union[EmailDeliveryIngredient[TypeEmailVerificationEmailDeliveryInput], None] = None):
+                 ingredients: Union[EmailVerificationIngredients, None] = None
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailverification/recipe.py:56]: Don't pass `None` as default value here

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 2]
```diff
@@ -236,11 +236,14 @@ def __init__(self, id: str, email: str):
 
 class TypeEmailVerificationEmailDeliveryInput:
     def __init__(
-        self, user: TypeEmailVerificationEmailDeliveryInputUser,
-        email_verify_link: str, user_context: Dict[str, Any]
+        self,
+        user: TypeEmailVerificationEmailDeliveryInputUser,
+        email_verify_link: str,
+        user_context: Dict[str, Any]
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailverification/interfaces.py:242]: send_email function takes email_input and user_context as parameter. I think you can reomve user_context from here

File: supertokens_python/recipe/emailpassword/recipe.py [hunk 3]
```diff
@@ -88,9 +90,10 @@ def __init__(self, recipe_id: str, app_info: AppInfo,
         if email_verification_recipe is not None:
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/recipe.py:90]: make ingredients changes for __init__ function here too similar to the email verification recipe

File: supertokens_python/recipe/emailpassword/recipe.py [hunk 4]
```diff
@@ -12,19 +12,21 @@
...
 from supertokens_python.recipe.emailpassword.types import \
     TypeEmailPasswordEmailDeliveryInput
-from supertokens_python.recipe.emailverification.interfaces import TypeEmailVerificationEmailDeliveryInput
+from supertokens_python.recipe.emailverification.interfaces import \
+    TypeEmailVerificationEmailDeliveryInput
+from supertokens_python.recipe.emailverification.types import \
+    EmailVerificationIngredients
 from supertokens_python.recipe_module import APIHandled, RecipeModule
 
 from ...exceptions import SuperTokensError
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/recipe.py:32]: in the `__init__` function below, make ingredients changes similar to email_verification recipe

File: supertokens_python/recipe/emailverification/emaildelivery/services/backwardCompatibility/__init__.py [hunk 5]
```diff
@@ -1,15 +1,39 @@
...
-from supertokens_python.recipe.emailverification.utils import \
-    default_create_and_send_custom_email
 from supertokens_python.supertokens import AppInfo
+from httpx import AsyncClient
+from os import environ
+
+
+def default_create_and_send_custom_email(app_info: AppInfo) -> Callable[[User, str, Dict[str, Any]], Awaitable[None]]:
+    async def func(user: User, email_verification_url: str, _: Dict[str, Any]):
+        if ('SUPERTOKENS_ENV' not in environ) or (environ['SUPERTOKENS_ENV'] != 'testing'):
```
Comment (KShivendu) [inline supertokens_python/recipe/emailverification/emaildelivery/services/backwardCompatibility/__init__.py:29]: This was previously `('SUPERTOKENS_ENV' not in environ) or (environ['SUPERTOKENS_ENV'] != 'testing')` which is the negation of the current condition. Fixed it.

File: supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py [hunk 6]
```diff
@@ -32,53 +31,84 @@
...
+                 ) -> None:
         self.app_info = app_info
         self.recipeInterfaceImpl = recipeInterfaceImpl
-        self.reset_password_using_token_feature = reset_password_using_token_feature
 
-        if reset_password_using_token_feature:
-            if reset_password_using_token_feature.create_and_send_custom_email is None:
-                reset_password_using_token_feature.create_and_send_custom_email = default_create_and_send_custom_email(self.app_info)
-            self.reset_password_using_token_feature = reset_password_using_token_feature
+        rputf_create_and_send_custom_email = default_create_and_send_custom_email(self.app_info)
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py:84]: change rputf_create_and_send_custom_email to reset_password_feature_send_email_func

File: supertokens_python/recipe/emailpassword/types.py [hunk 7]
```diff
@@ -84,3 +87,10 @@ def __init__(
     TypeEmailPasswordPasswordResetEmailDeliveryInput,
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/types.py:87]: Similar to `TypeEmailVerificationEmailDeliveryInput`, you don't require user_context in `TypeEmailPasswordPasswordResetEmailDeliveryInput`

File: supertokens_python/recipe/emailpassword/types.py [hunk 8]
```diff
@@ -84,3 +87,10 @@ def __init__(
...
     TypeEmailVerificationEmailDeliveryInput
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/types.py:88]: Similar to `TypeEmailVerificationEmailDeliveryInput`, you don't require user_context in `TypeEmailPasswordPasswordResetEmailDeliveryInput`

Comment (KShivendu) [reply supertokens_python/recipe/emailpassword/types.py:88]: @bhumilsarvaiya If I do that, https://github.com/supertokens/supertokens-python/pull/102/files#diff-074864472d77ea0c5ef0cd85a5b5bedddd0c510b6c7a0aa70cafd20017331ad0R131 starts saying that it needs UserContext. The `user_context` in `get_user_by_id` isn't getting used but it's present. So need to pass something. 

We can pass `{}` but won't that be dangerous?

File: supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py (hunk 6 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py:84]: Much better. Thanks!

File: supertokens_python/recipe/emailpassword/types.py (hunk 8 above)
Comment (bhumilsarvaiya) [reply supertokens_python/recipe/emailpassword/types.py:88]: we already have `user_context` passed as parameter for `send_email`. Pass that to `get_user_by_id` instead

File: supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py [hunk 9]
```diff
@@ -98,7 +128,7 @@ async def send_email(self, email_input: TypeEmailPasswordEmailDeliveryInput, use
 
             try:
                 await self.reset_password_using_token_feature.create_and_send_custom_email(
-                   user, email_input.password_reset_link, email_input.user_context
+                    user, email_input.password_reset_link, email_input.user_context
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py:131]: `email_input.user_context` --> `user_context`

File: supertokens_python/recipe/emailverification/emaildelivery/service/backwardCompatibility/__init__.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+            return
+        try:
+            async with AsyncClient() as client:
+                await client.post('https://api.supertokens.io/0/st/auth/email/verify', json={'email': user.email, 'appName': app_info.app_name, 'emailVerifyURL': email_verification_url}, headers={'api-version': '0'})  # type: ignore
+        except Exception:
+            pass
+    return func
+
+
+class CreateAndSendCustomEmailInput:
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailverification/emaildelivery/service/backwardCompatibility/__init__.py:39]: remove this if not used

File: supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py [hunk 11]
```diff
@@ -32,53 +31,84 @@
...
+            }
+            async with AsyncClient() as client:
+                await client.post('https://api.supertokens.io/0/st/auth/password/reset', json=data, headers={'api-version': '0'})  # type: ignore
+        except Exception:
+            pass
+
+    return func
+
 
 class CreateAndSendCustomEmailInput:
```
Comment (bhumilsarvaiya) [inline supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py:60]: remove this if not used

Comment (KShivendu) [reply supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py:60]: done.

File: supertokens_python/recipe/emailverification/emaildelivery/service/backwardCompatibility/__init__.py (hunk 10 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/emaildelivery/service/backwardCompatibility/__init__.py:39]: done.

File: supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py (hunk 9 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailpassword/emaildelivery/service/backwardCompatibility/__init__.py:125]: Done. Thanks!

**Pull request merged**